Dusty live artifacts

dustymabe opened the pull request.

jlebon committed:
stages: add `coreos.live-artifacts.mono` stage

This adds a new `org.osbuild.coreos.live-artifacts.mono` stage to build
CoreOS Live ISO/PXE artifacts. The code is heavily based on the
`cmd-buildextend-live` script from coreos-assembler [1], but a lot of
things had to be adapted:
- the stage is provided the deployed oscontainer tree, metal, and
  metal4k images as inputs
- we use chroot instead of supermin to execute some commands in the
  context of the target oscontainer
- a bunch of calls that were wrapped by libguestfs for us (e.g.
  mkfs.vfat, mksquashfs), we now have to call ourselves; to retain
  maximum compatibility, we ensured that we still effectively use the
  same args that libguestfs passed

And various other minor adjustments.

Of course, this is not really in line with the OSBuild philosophy
of having smaller-scoped stages. We have labeled this with a .mono
suffix to denote it is monolithic, similar to the existing
`org.osbuild.bootiso.mono` stage today.

Eventually we may be able to break this stage down if we find it worth
the effort. Alternatively the need for it may go away as we align more
with Image Mode.

[1] https://github.com/coreos/coreos-assembler/blob/43a9c80e1f548269d71d6d586f0d5754c60f6144/src/cmd-buildextend-live

Co-authored-by: Dusty Mabe <dusty@dustymabe.com>
Co-authored-by: Renata Ravanelli <renata.ravanelli@gmail.com>

ravanelli committed:
util/chroot: Add support for custom directory bind mounts

 - Add optional bind_mounts parameter to __init__ method;
 - Enhanced methods to accept an optional `bind_mounts`.
This allows for more flexible for configurations when setting
up bind mounts.

Signed-off-by: Renata Ravanelli <rravanel@redhat.com>

ravanelli committed:
CODEOWNERS: Assign ownership for CoreOS-related files

- Introduce a CODEOWNERS file to designate the CoreOS team as responsible for maintaining files related to CoreOS.
- This change ensures collaborative support, alleviating the burden solely on the osbuild team.
- By assigning ownership, we aim to enhance code quality and responsiveness to issues affecting CoreOS.

Signed-off-by: Renata Ravanelli <rravanel@redhat.com>

thozza committed:
Testutil: add DNF5 helper function for depsolving

This will be used for testing the SBOM implementation with DNF5.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
test_util_sbom_dnf: fix imported module name

There was a typo in the imported module name, which caused the test to
be always skipped.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
test_util_sbom_spdx: fix imported module name

There was a typo in the imported module name, which caused the test to
be always skipped.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
util/sbom: add support for DNF5

Add functions for transforming package sets depsolved using libdnf5 to
the SBOM standard-agnostic model. Cover the function with unit tests.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
osbuild-depsolve-dnf: add SBOM support for DNF5

Enable generating SBOM documents for depsolved transactions when using
DNF5. Enable SBOM testing with DNF5 in unit tests.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
util/sbom/spdx: rename {,s}bom_pkgset_to_spdx2_doc()

Rename the function for consistency reason. The parent package is named
SBOM (originally BOM).

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

ravanelli committed:
util/osrelease.py:  Replaced string stripping with `shlex.split()`

- Replaced string stripping with `shlex.split()` to properly
handle values in the os-release file;
- This ensures cleaner and more accurate key-value assignments,
follwing a broader set of shell-like parsing rules;
- Add os-release file for Fedora CoreOS 40 for testing.

Signed-off-by: Renata Ravanelli <rravanel@redhat.com>

thozza committed:
Schutzfile: use released repositories for F41

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
Test/manifests: update fedora manifests to F41

Update the fedora manifest template to F41. This is a preparation for
testing a new `org.osbuild.dnf5.sbom.spdx` stage. In addition to that,
F39 is already EOL and unsupported, so moving to a newer version is
desirable.

Regenerate all testing manifests.

Notes about specific changes:
 - Remove `pcmciautils` package from ostree manifest, because it has
   been deprecated and it is not available on F41 any more.
 - Add `python3-dnf` package to the Fedora build pipeline. The reason
   is that DNF5 is installed by default since F41 and the module is
   needed to test e.g. `org.osbuild.dnf4.sbom.spdx` or
   `org.osbuild.dnf4.mark` stages.
 - Add `python3-libdnf5` package to the Fedora build pipeline. This is a
   preparation for testing a new `org.osbuild.dnf5.sbom.spdx` stage.
 - Install `pam` in `authselect` stage test manifest, because otherwise
   authselect fails on the nonexistence of /etc/pam.d.
 - Adjust the `dnf4.mark` test case to also accept `dnf5` as the marked
   package, because on F41, `dnf` is a virtual provide of it.
 - Adjust stage tests and their diffs as needed.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
Add org.osbuild.dnf5.sbom.spdx stage

Add DNF5 version of the `sbom.spdx` stage, which generates an SPDX SBOM
document from a given FS tree using libdnf5 API.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

alexlarsson committed:
osbuild-mpp: Fix error on python < 3.11

The change in commit ed3386943069e138b346eef75ac64d9874a07810 to
use hashlib.file_digest breaks with older python, because
it was added in 3.11.

This change reverts back to hashing all the data in the case
where file_digest doesn't exist.

dustymabe committed:
manifests: updates for FCOS manifest

There have been a lot of changes to the CoreOS definitions in [1].
Let's update the test manifest here to more closely match what is
running in the field there.

[1] https://github.com/coreos/coreos-assembler/tree/dcd60cfe01f1d42e8d5f63a77db8172f2f1101ae/src/osbuild-manifests

thozza committed:
Tests/tox: latest Fedora uses Python 3.13, so use it

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
GH actions: use the latest build of osbuild-ci* containers

This will bring the latest image builds with new dependencies for unit
tests.

In addition, pin down the osbuild-ci-c*s container to a specific tag, to
ensure stability of the CI.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

achilleas-k committed:
stages: use util.chroot in all stages that call "chroot"

Use the chroot utility module for all cases where we need to chroot
during a stage's execution.

The advantage is that all stages use the same tested code path for
setting up a chroot and all chrooted commands run in the same
environment, with the /proc, /dev, and /sys filesystems mounted.

dustymabe committed:
osbuild/remoteloop: add more loop device options

This adds lock, partscan, read_only, sector_size to _create_device()
similar to make_loop() from devices/org.osbuild.loopback.

mvo5 committed:
sources: fix ostree_sources test to work without `/var/empty`

Not all distros ship `/var/empty` so just create an empty dir
on demand as needed.

This also tweaks `test_ostree_source_exists()` into calling
`make_repo()` instead of duplicating that code.

mvo5 committed:
workflow: add smoke test that runs unit tests as user

This commit adds a tiny smoke test that runs the unit tests as
a normal user on a regular github runner. This should ensure that
we catch issues like in osbuild#1936.

mvo5 committed:
workflow: install `python3-pytest` too to workaround `test_host.py`

The test_host.py:test_signals_on_separate_fd (and more)  runs
itself but that run will happen without the tox env so a pip/tox
installed pytest will not be found, install the pytest package
as a workaround.

achilleas-k committed:
test/users: modify root's home directory in stage test

Since 9071cd0abbf8db059aa34e6cc21612da244e0684 the commands in the
org.osbuild.users stage are run using the osbuild.util.chroot.Chroot
context manager.  This includes mounting /proc into the chroot.

The `usermod` command now fails with

  usermod: user root is currently used by process N

when modifying any property of the root user that requires the user to
not be logged in (like the home directory).

Change the home directory of the root user in the stage test to check
for this issue and catch regressions.

achilleas-k committed:
stages: run useradd, usermod, and passwd with --root

The useradd, usermod, and passwd commands support a `--root` option that
handles chroot-ing for the command.  In general, we prefer using this
option for commands that provide it and relying on the utility itself to
know how to set up the chroot in the way it needs.

The option has been available for these commands since 2011 [1] and it's
unclear why they weren't used originally.

The `mkhomedir_helper` command is still run using our Chroot context, so
the fix introduced in 9071cd0abbf8db059aa34e6cc21612da244e0684 is
unaffected.

[1] https://github.com/shadow-maint/shadow/blob/365279ea95a6c76021e2100e51e71b1991fea32b/ChangeLog#L1339

achilleas-k committed:
test: update users stage unit tests

Update the unit tests to match new calls and options.

mvo5 committed:
sources: skip ostree tests if no ostree binary if found

This commit skips the ostree tests if no ostree binary is available.

cgwalters commented:
>  Alternatively the need for it may go away as we align more with Image Mode.

Yes, I think the debate to have here is whether it really continues to make sense to have "dd raw image" installations versus having coreos-installer grow small bits to dynamically make the desired filesystems and then call `bootc install to-filesystem`. (And actually nowadays, `bootc install to-disk` acts as such a reference; I consider it more of a "demo" and testing tool but it's basically that same code)

If we don't do the raw image for installs but leverage `bootc install`, then the center of gravity shifts much more decisively to containers (and bootc), and about half of the magic in the ISO drops out.

jlebon commented on osbuild/loop.py:
```diff
@@ -126,6 +126,8 @@ def __init__(self, minor, dir_fd=None):
             if not dir_fd:
                 dir_fd = os.open("/dev", os.O_DIRECTORY)
                 stack.callback(lambda: os.close(dir_fd))
+            if not os.path.exists(os.path.join("/dev", self.devname)):
```
Hmm, the dfd-relative-ness of the surrounding code makes it seem like it's possible for `dir_fd` to not refer to `/dev`. So we should probably use a dfd-relative API here too. E.g.

```python
try:
    os.stat(self.devname, dfd=dir_fd)
except FileNotFound:
    self.mknod(dir_fd)
```

dustymabe replied:
done

jlebon commented on stages/org.osbuild.coreos.live-artifacts.mono.meta.json:
```diff
@@ -0,0 +1,68 @@
+{
+  "summary": "Build CoreOS Live ISO and PXE kernel,initramfs,rootfs",
+  "description": [
+    "This stage builds the CoreOS Live ISO and PXE kernel,initramfs,rootfs",
```
Minor: hmm, is there a reason for no spacing in the commas?

dustymabe replied:
just that they are all PXE related.. would `kernel/initramfs/rootfs` be better?

jlebon replied:
How about

> This stage builds the CoreOS live ISO and PXE artifacts (kernel, initramfs, and rootfs)

?

jlebon commented on stages/org.osbuild.coreos.live-artifacts.mono:
```diff
@@ -0,0 +1,841 @@
+# on the provided metal and metal4k disk images that are provided as
+# inputs.
+#
+# For historical context and to see its evolution, refer to the original source:
```
s/evolution/git history/  ?

jlebon commented on stages/org.osbuild.coreos.live-artifacts.mono:
```diff
@@ -0,0 +1,841 @@
+# 3. live-initramfs (i.e. images/pxeboot/initrd.img)
+# 4. live-rootfs    (i.e. images/pxeboot/rootfs.img)
+#
+# Thie live-iso itself contains the other 3 artifacts inside of
```
This

dustymabe replied:
I think I meant `the`.

jlebon commented on stages/org.osbuild.coreos.live-artifacts.mono:
```diff
@@ -0,0 +1,841 @@
+    # For dev/test purposes it is useful to create test fixture ISO
+    # images for the coreos-installer CI framework to use [1]. This test
+    # fixture is much smaller than an actual ISO and can be stored in git.
+    # Determine if the user want's a test fixture ISO created by checking
```
wants

jlebon commented on stages/org.osbuild.coreos.live-artifacts.mono:
```diff
@@ -0,0 +1,841 @@
+        os.makedirs(tmpinitrd_rootfs, exist_ok=True)
+
+        # Remove the sysroot=readonly flag, see https://github.com/coreos/fedora-coreos-tracker/issues/589
+        subprocess.check_call(['sed', '-i', '/readonly=true/d', f'{tmp_squashfs_dir}/ostree/repo/config'])
```
Minor: add space here

dustymabe replied:
want to make a suggestion?

jlebon reviewed (approved):
Really just minor/superficial things, this LGTM overall! Thanks for adding all the comments.

jlebon commented:
> > Alternatively the need for it may go away as we align more with Image Mode.
> 
> Yes, I think the debate to have here is whether it really continues to make sense to have "dd raw image" installations versus having coreos-installer grow small bits to dynamically make the desired filesystems and then call `bootc install to-filesystem`. (And actually nowadays, `bootc install to-disk` acts as such a reference; I consider it more of a "demo" and testing tool but it's basically that same code)
> 
> If we don't do the raw image for installs but leverage `bootc install`, then the center of gravity shifts much more decisively to containers (and bootc), and about half of the magic in the ISO drops out.

I think before we can move away from dd and osmet, we probably want much better sharing between the live host and container storage in image mode (which probably means composefs).

Higher-level, I've been thinking of convergence in that area more as "build the CoreOS live ISO using shared image mode tooling" which involves adding the features we need there and is more like replacing rather than adapting. Though I'm not sure how realistic that is. Definitely, this needs more thought and honestly needs its own OpenShift enhancement to surface all the requirements there.

But anyway, I don't think this is worth tackling in the short to mid-term, and there's lots of other places where more useful convergence work needs to happen first. But yes, it should be on the roadmap.

jlebon commented on stages/org.osbuild.coreos.live-artifacts.mono:
```diff
@@ -0,0 +1,841 @@
+
+        # Remove the sysroot=readonly flag, see https://github.com/coreos/fedora-coreos-tracker/issues/589
+        subprocess.check_call(['sed', '-i', '/readonly=true/d', f'{tmp_squashfs_dir}/ostree/repo/config'])
+        # And ensure that the kernel binary and hmac file is in the place that dracut
```
```suggestion

        # And ensure that the kernel binary and hmac file is in the place that dracut
```

dustymabe commented:
I addressed code review comments and also fixed lints that would run in upstream CI.

dustymabe committed:
test/data/manifests/fedora-coreos-container: fix CI

We updated the container in the registry so we need to update this
manifest to unbreak CI.

tkoscieln committed:
Bump RHEL 10 from beta to nightly

Bumping Rhel 10 runners from beta to nightly, running manifests on 9.6 nightly and also removing the Rhel 9.5 nightly runners altogether.

pypingou committed:
Expand the options available in containers.unit.create

Add the option to specify the SELinux type label and file type label for
the process.
Add the option to specify a tmpfs mount to create into the container.

Signed-off-by: Pierre-Yves Chibon <pingou@pingoured.fr>

achilleas-k committed:
test: add new containers.unit.create options in test

ondrejbudai committed:
github: prevent script injections via PR branch names

Prior this commit, ${{ github.event.workflow_run.head_branch }} got
expanded in the bash script. A malicious actor could inject
an arbitrary shell script. Since this action has access to a token
with write rights the malicious actor can easily steal this token.

This commit moves the expansion into an env block where such an
injection cannot happen. This is the preferred way according to the
github docs:
https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable

schutzbot committed:
Post release version bump

[skip ci]

mvo5 committed:
util,test: add test for new shlex based os-release parsing

This commit adds a tiny unit test for the new `shlex` based
os-release parsing and tweaks the error message in a small
and non-functional way (just because it's slightly nicer
for a user). The test checks for three keys NAME which is
quoted with `"`, ID which is not quoted and OSTREE_VERSION
which is quoted with `'`.

brlane-rht committed:
org.osbuild.grub2.iso.legacy: Add grub2 setup for booting BIOS ISO

This is the BIOS version of the grub2 iso stage. It installs the config
file and copies over the grub2 modules to /boot/grub2/

dustymabe committed:
test: add podman to fedora-build-v2 manifest

The org.osbuild.container-deploy stage uses podman. Including it
in the build here will allow that stage to be used with this
pipeline as the buildroot.

Include a workaround here for what I consider to be a bug [1] in that
`podman` will create `/etc/containers/networks` on first run if it
doesn't exist. That dir should just be created by an RPM. If we
don't include this workaround then the stage will fail when `podman`
attempts the `mkdir` because `/etc/containers` is mounted in from
the buildroot readonly.

[1] https://github.com/containers/common/pull/2265

mvo5 committed:
stage: add test for core live-artifacts stage

This commit adds some unit tests around the coreos live-artifcats
mono stage.

- test/coreos_live_artifacts: add test for align_initrd()
- test/coreos_live_artifacts: add test for extend_initramfs()
    - This actually tests the mkinitrd_pipe() function, which
      extend_initramfs() calls after opening the file.
- test/coreos_live_artifacts: add test for make_stream_hash()
- test/coreos_live_artifacts: add test for make_efi_bootfile()

Co-authored-by: Achilleas Koutsou <achilleas@koutsou.net>

brlane-rht committed:
typo: Fix incorrect 'tar archive' descriptions in stage metadata

SSIA

schutzbot committed:
schutzfile: Update snapshots to 20241210

achilleas-k committed:
stages/coreos.live-artifacts: refactor into more functions

This improves the readability and maintainability of the code to
have it split into smaller pieces.

Co-Authored-by: Dusty Mabe <dusty@dustymabe.com>

ochosi committed:
github/actions: Enable /jira-epic slash commands

This change allows for using the command to create Jira Tasks under a
given Epic both in a pull request comment or in the pull request
description.

dustymabe committed:
stages/coreos.live-artifacts: drop kernel rename for s390x

The rename to get file names + suffix with < 8 characters
was done in [1] when our initramfs was named initramfs.img.
It was subsequently renamed to initrd.img in [2] and the
rename of the initramfs was dropped but the rename of the
kernel was never dropped. Since vmlinuz is already < 8
characters let's just drop the rename here too.

[1] https://github.com/coreos/coreos-assembler/commit/60400910bb2e91a1ac6c01819b9062302e050e5a
[2] https://github.com/coreos/coreos-assembler/commit/6f533ef55ee5cf2b10ad97df51eb2d190b6b2c2b

dustymabe committed:
stages/coreos.live-artifacts: drop usage of dir fd

Rework rename of vendor directory to not use dfd APIs.

This was requested in code review.

Also added comments since I now understand it better.

dustymabe committed:
stages/coreos.live-artifacts: stop chmod initrd.img

The initramfs.img in Fedora and RHEL CoreOS images is already 644
so this isn't needed.

dustymabe committed:
stages/coreos.live-artifacts: move rename of vendor dir

Seems more appropriate in copy_configs_and_init_kargs_json() since this
is where files were originally copied in.

thozza committed:
tools: add image-info tool as osbuild-image-info

Add the image-info tool as osbuild-image-info from the manifest-db
repository [0]. This is an exact copy without any changes.

[0] https://github.com/osbuild/manifest-db/blob/8e05a898d4a83f31ad24496c8738a3b1724cf937/tools/image-info

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
setup.py: install tools/osbuild-image-info

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
SPEC: include osbuild-image-info in the -tools subpackage

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
tools/osbuild-image-info: fix issues reported by autopep8

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
CI: move the manifest_tests.sh from schutzbot/ to test/cases/

The test script should have been in the test/cases/ directory in the
first place.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
SELinux: apply osbuild_exec_t to /usr/bin/osbuild-image-info

This is needed, for the tool to be able to read SELinux labels from the
inspected image, which are not known to the host on which it is running.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>

thozza committed:
manifest_tests.py: use osbuild-image-info from osbuild package

Modify the test case to not use the image-info copy from manifest-db.

Signed-off-by: Tomáš Hozza <thozza@redhat.com>
